### PR TITLE
[air] New `train.Checkpoint` API: Update `Train tests + examples` (batch 1)

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

This PR updates the first batch of `Train tests + examples` CI tests to use the new persistence mode and checkpoint API. Each converted test will add the `new_storage` tag in the CI, which excludes it from the old runners. This is so that we don't need to keep backwards compatibility within the tests.

## Related issue number

https://github.com/ray-project/ray/issues/38570

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/experiment/trial.py [hunk 1]
```diff
@@ -1093,6 +1093,7 @@ def on_checkpoint(self, checkpoint: Union[_TrackedCheckpoint, _TrainingResult]):
             self.storage.current_checkpoint_index += 1
         else:
             self.run_metadata.checkpoint_manager.on_checkpoint(checkpoint)
+        self.invalidate_json_state()
```
Comment (justinvyu) [inline python/ray/tune/experiment/trial.py:1096]: `self.storage` is not part of `run_metadata`, so it's not enough to just invalidate the `run_metadata` cache. @krfricke

File: python/ray/train/tests/test_torch_trainer.py [hunk 2]
```diff
@@ -367,8 +294,10 @@ def __getstate__(self):
         model = CustomLinear(1, 1)
         model = train.torch.prepare_model(model)
 
-        # Make sure model is serializable even with amp enabled.
-        train.report({}, checkpoint=LegacyTorchCheckpoint.from_model(model))
+        # TorchCheckpoint.from_model fails, so just save the state dict only.
+        train.report(
+            {}, checkpoint=TorchCheckpoint.from_state_dict(model.module.state_dict())
+        )
```
Comment (justinvyu) [inline python/ray/train/tests/test_torch_trainer.py:300]: @amogkam This test fails if I try to `torch.save(model)` or `pickle.dump(model)`. Is the fix from https://github.com/ray-project/ray/pull/25335 supposed to fix that?

Is this unit test still relevant? Saving the state dict seems to work fine.

File: python/ray/train/tests/test_trainer_restore.py [hunk 3]
```diff
@@ -161,30 +174,34 @@ def test_gbdt_trainer_restore(ray_start_6_cpus, tmpdir, trainer_cls):
...
 @pytest.mark.parametrize("trainer_cls", [TransformersTrainer])
-def test_trainer_with_init_fn_restore(ray_start_4_cpus, tmpdir, trainer_cls):
+def test_trainer_with_init_fn_restore(
+    ray_start_4_cpus, tmp_path, trainer_cls, monkeypatch
+):
     """Tests restore for data parallel trainers that take in a `train_init` function
     and config. Success criteria: same as for data parallel trainers."""
+    # TODO(justinvyu): Failure injection callback doesn't work very well to test this.
+    pytest.skip("Re-enable after coming up with a better way to inject a failure.")
+    monkeypatch.setenv("RAY_AIR_LOCAL_CACHE_DIR", str(tmp_path))
```
Comment (justinvyu) [inline python/ray/train/tests/test_trainer_restore.py:204]: This test is failing for some complicated reasons:
1. We inject a failure in a callback. Errors in callbacks get handled in different ways.
2. If we fail during `on_trial_save`, the error gets raised immediately WITHOUT calling `tune_controller.checkpoint` one last time. This means that the checkpoint gets processed ->  checkpoint 00001 gets deleted, but the experiment state still thinks checkpoint 00001 is the latest one. This is a consequence of raising exactly during `on_trial_save` + `num_to_keep=1`.
3. I tried failing on the next iteration's `on_trial_result`, but that also didn't work. The trial advanced to the next iteration for some reason.

So, I'm skipping for these (legacy) trainers for now.

File: python/ray/train/tests/test_gpu_examples.py [hunk 4]
```diff
@@ -98,6 +98,12 @@ def test_tune_tensorflow_mnist_gpu(ray_start_4_cpus_2_gpus):
 
 
 def test_train_linear_dataset_gpu(ray_start_4_cpus_2_gpus):
+    from ray.train._internal.storage import _use_storage_context
+
+    if not _use_storage_context():
+        # TODO(justinvyu): [skipped_test]
+        pytest.skip("Skipping for now.")
```
Comment (justinvyu) [inline python/ray/train/tests/test_gpu_examples.py:105]: This is in a different (GPU) test suite, but shares example code for a different test that has been updated, so excluding it here.

File: python/ray/train/tests/test_trainer_restore.py (hunk 3 above)
Comment (krfricke) [reply python/ray/train/tests/test_trainer_restore.py:204]: Fine with me to skip the test, but for the case where `on_trial_save` fails (in a custom callback or so), should we make sure this works? (Can be in a follow-up)

**Review by krfricke — APPROVED** [review]
Test changes lgtm

File: python/ray/train/tests/test_trainer_restore.py (hunk 3 above)
Comment (justinvyu) [reply python/ray/train/tests/test_trainer_restore.py:204]: Sounds good. I'm not sure why this worked before actually, would be good to get to the bottom of this in a follow-up.

Tracking that here https://github.com/ray-project/ray/issues/38734

**Pull request merged**